google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractMultiUseTransaction.java

```diff
@@ -17,22 +17,71 @@
 package com.google.cloud.spanner.connection;
 
 import com.google.api.core.ApiFuture;
+import com.google.cloud.spanner.Dialect;
 import com.google.cloud.spanner.ErrorCode;
 import com.google.cloud.spanner.Options.QueryOption;
 import com.google.cloud.spanner.ReadContext;
 import com.google.cloud.spanner.ResultSet;
 import com.google.cloud.spanner.SpannerException;
 import com.google.cloud.spanner.SpannerExceptionFactory;
 import com.google.cloud.spanner.connection.AbstractStatementParser.ParsedStatement;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 import com.google.spanner.v1.SpannerGrpc;
+import java.util.LinkedList;
+import java.util.Objects;
+import javax.annotation.Nonnull;
 
 /**
  * Base class for {@link Connection}-based transactions that can be used for multiple read and
  * read/write statements.
  */
 abstract class AbstractMultiUseTransaction extends AbstractBaseUnitOfWork {
 
+  /** In-memory savepoint implementation that is used by the Connection API. */
+  static class Savepoint {
+    private final String name;
+
+    static Savepoint of(String name) {
+      return new Savepoint(name);
+    }
+
+    Savepoint(String name) {
+      this.name = name;
+    }
+
+    /** Returns the index of the first statement that was executed after this savepoint. */
+    int getStatementPosition() {
+      return -1;
+    }
+
+    /** Returns the index of the first mutation that was executed after this savepoint. */
+    int getMutationPosition() {
+      return -1;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (!(o instanceof Savepoint)) {
+        return false;
+      }
+      return Objects.equals(((Savepoint) o).name, name);
+    }
+
+    @Override
+    public int hashCode() {
+      return name.hashCode();
+    }
+
+    @Override
+    public String toString() {
+      return name;
+    }
+  }
+
+  private final LinkedList<Savepoint> savepoints = new LinkedList<>();
```

Comment:
I considered a couple of different options, but landed on `LinkedList` both for simplicity, but also because it aligns with how `SAVEPOINT` works. They are strictly ordered by an index (i.e. a map structure makes less sense, as there is no mapping between a key and a value other than a simple index key), and removing a savepoint from the list always means removing everything after it as well. That means that cutting one node in the middle of the list is a cheap operation.
That being said; a transaction is unlikely to ever have more than a handful of savepoints, meaning that the chosen data structure won't have any practical effect on the performance.

- `Deque` was indeed one that I considered, but I dropped it as we need to be able to peek multiple steps back, which is not supported by a pure `Deque` implementation.
- `TreeMap` or similar would not really be a good fit, as the key in this case is just an index. The savepoint names also do not need to be unique (in PostgreSQL).